pkg/distributors/contracts/MultiRewards.sol

```diff
@@ -390,245 +427,233 @@ contract MultiRewards is IMultiRewards, IDistributor, ReentrancyGuard, MultiRewa
+            address distributionToken = address(distribution.distributionToken);
 
-        distribution.duration = duration;
-        emit RewardDurationSet(distributionId, duration);
-    }
+            if (unclaimedTokens > 0) {
+                userDistribution.unclaimedTokens = 0;
+                emit TokensClaimed(msg.sender, distributionToken, unclaimedTokens);
+            }
 
-    /**
-     * @notice Updates the reward per staked token rate of a distribution
-     */
-    function _updateDistributionRate(bytes32 distributionId) internal returns (uint256 rewardPerTokenStored) {
-        Distribution storage distribution = _getDistribution(distributionId);
-        rewardPerTokenStored = _rewardPerToken(distribution);
-        distribution.rewardPerTokenStored = rewardPerTokenStored;
-        distribution.lastUpdateTime = _lastTimeRewardApplicable(distribution);
-    }
+            ops[i] = IVault.UserBalanceOp({
+                asset: IAsset(distributionToken),
+                amount: unclaimedTokens,
+                sender: address(this),
+                recipient: payable(recipient),
+                kind: kind
+            });
+        }
 
-    /**
-     * @notice Updates unpaid rewards due to a user
-     */
-    function _updateUserRewardRatePerToken(UserStaking storage userStaking, bytes32 distributionId) internal {
-        uint256 rewardPerTokenStored = _updateDistributionRate(distributionId);
-        UserDistribution storage userDistribution = userStaking.distributions[distributionId];
-        userDistribution.unpaidRewards = _totalEarned(userStaking, distributionId);
-        userDistribution.paidRatePerToken = rewardPerTokenStored;
+        getVault().manageUserBalance(ops);
     }
 
-    function _updateDistribution(IERC20 stakingToken, address user) internal {
+    function _updateDistributions(IERC20 stakingToken, address user) internal {
         UserStaking storage userStaking = _userStakings[stakingToken][user];
-        EnumerableSet.Bytes32Set storage distributions = userStaking.allowedDistributions;
+        EnumerableSet.Bytes32Set storage distributions = userStaking.subscribedDistributions;
         uint256 distributionsLength = distributions.length();
 
         for (uint256 i; i < distributionsLength; i++) {
             bytes32 distributionId = distributions.unchecked_at(i);
-            _updateUserRewardRatePerToken(userStaking, distributionId);
+            _updateUserTokensPerStake(userStaking, distributionId);
         }
     }
 
-    /* ========== MODIFIERS ========
+    function _updateUserTokensPerStake(UserStaking storage userStaking, bytes32 distributionId) internal {
+        uint256 globalTokensPerStake = _updateDistributionRate(distributionId);
+        UserDistribution storage userDistribution = userStaking.distributions[distributionId];
+        userDistribution.unclaimedTokens = _totalUnclaimedTokens(userStaking, distributionId);
```

Comment:
This is not incorrect, but calling `_totalUnclaimedTokens` here is quite inefficient as most of what it does is recompute things we've already computed (e.g. check that the user is subscribed, compute globalTokensPerStake, etc.).

Reply:
Agreed, we can pass down some more storage pointers but in a later PR.

Reply:
Created https://github.com/balancer-labs/balancer-v2-monorepo/issues/1013